### DISCUSSION TITLE
Accessing secrets by index using an environment variable

### DISCUSSION THREAD

**Discussion by nilsdebruin** [community/community#25171] [Actions]
<p>Hi all!</p>
<p>At this moment I am creating a pipeline in which I would like to achieve the following:</p>
<ul>
<li>
<p>Set an environment variable based on a certain condition</p>
</li>
<li>
<p>Use that environment variable to access a certain secret which I have defined (in this case PROD)</p>
</li>
</ul>
<p>To make this more clear, I added a code snippet:</p>
<pre><code>-name: Set env to production
if: endsWith(github.ref, '/master')
run: |
echo "::set-env name=ENVIRONMENT::PROD"

- run: echo ${{ secrets[$ENVIRONMENT] }}
</code></pre>
<p>So based on the set environment variable I would like to access: secrets.PROD</p>
<p>I used: <a href="https://help.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions" rel="nofollow noopener">https://help.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions</a> as a reference. Any help or direction would be appreciated!</p>

**Comment by jdbruijn** [comment]
<p>I’m not sure whether that is possible. I use something similar in one of my workflows (<a href="https://github.com/vidavidorra/repo-template/blob/7210f3746e818037cd58f4ed4c0c4cb032410b4c/.github/workflows/release.yml#L36" rel="nofollow noopener">see here</a>).</p>
<p>I believe environments variables are referenced in a different way that you’re doing since this is about environment variables between different steps. You should use “env.ENVIRONMENT” to reference an environment variable.</p>
<p>The following example should echo “PROD”</p>
<pre><code>- name: Set env to production
  if: endsWith(github.ref, '/master')
  run: |
    echo "::set-env name=ENVIRONMENT::PROD"
- run: echo ${{ env.ENVIRONMENT }}
</code></pre>
<p>Note that you could use bash in a step to selectively get an environmnet variable (also see my workfow for bash in a “run” step.</p>
<p>Hope this helps any.</p>

**Comment by Yanjingzhu** [comment, marked as answer]
<p>Hi <a class="mention" href="/nilsdebruin">@nilsdebruin</a>,  </p>
<p>Accessing secrets by index using an environment variable is not supported.<br>
As a workaround , you could add another step with if conditional for non-production . And you could use secrets.PROD directly when set a <strong>step environment</strong> variable. </p>
<pre><code>name: get secrets variable name from another env
on: push
jobs:
  get-secret:
    runs-on: ubuntu-latest
    steps:
    - name: Set env to production
      if: endsWith(github.ref, '/master')
      run: echo $environment
      env:
        environment: ${{secrets.PROD}}
    - name: Set env to non production
      if: "!endsWith(github.ref, '/master')"
      run: echo $environment
      env:
        environment: ${{secrets.TEST}}
</code></pre>

**Comment by nilsdebruin** [comment]
<p>Hi all,</p>
<p>Thanks for the replies! I have now come up with the following solution (which works, but does not feel dry):</p>
<pre><code>- name: Set env to staging
  if: endsWith(github.ref, '/develop')
  run: |
    echo "::set-env name=ENVIRONMENT::develop"
    echo "::set-env name=ENV_FILE::env.develop"
    echo "::set-env name=AWS_ACCESS_KEY_ID::${{ secrets.DEVELOP_AWS_ACCESS_KEY_ID }}"
    echo "::set-env name=AWS_SECRET_ACCESS_KEY::${{ secrets.DEVELOP_AWS_SECRET_ACCESS_KEY }}"
- name: Set env to production
  if: endsWith(github.ref, '/master')
  run: |
    echo "::set-env name=ENVIRONMENT::prod"
    echo "::set-env name=ENV_FILE::env.prod"
    echo "::set-env name=AWS_ACCESS_KEY_ID::${{ secrets.PROD_AWS_ACCESS_KEY_ID }}"
    echo "::set-env name=AWS_SECRET_ACCESS_KEY::${{ secrets.PROD_AWS_SECRET_ACCESS_KEY }}"

- name: Configure AWS Credentials
	uses: aws-actions/configure-aws-credentials@v1
	with:
	  aws-access-key-id: ${{ env.AWS_ACCESS_KEY_ID }}
	  aws-secret-access-key: ${{ env.AWS_SECRET_ACCESS_KEY }}
</code></pre>
<p>This works for my use case, so thanks for your help and input!</p>
